<a href="https://colab.research.google.com/github/varthini-30/tasksgenai/blob/main/Language_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
with open("/content/data (3).txt","r",encoding="utf-8") as f:
  data=f.read()
data

'DRACULA\n\nCHAPTER I\n\nJONATHAN HARKER\'S JOURNAL\n\n\n_3 May. Bistritz._--Left Munich at 8:35 P. M., on 1st May, arriving at\nVienna early next morning; should have arrived at 6:46, but train was an\nhour late. Buda-Pesth seems a wonderful place, from the glimpse which I\ngot of it from the train and the little I could walk through the\nstreets. I feared to go very far from the station, as we had arrived\nlate and would start as near the correct time as possible. The\nimpression I had was that we were leaving the West and entering the\nEast; the most western of splendid bridges over the Danube, which is\nhere of noble width and depth, took us among the traditions of Turkish\nrule.\n\nWe left in pretty good time, and came after nightfall to Klausenburgh.\nHere I stopped for the night at the Hotel Royale. I had for dinner, or\nrather supper, a chicken done up some way with red pepper, which was\nvery good but thirsty. (_Mem._, get recipe for Mina.) I asked the\nwaiter, and he said it 

RecurrentNurealNetwork,LSTM,GRU-NLP ,SEQUENTIAL MODEL ,SELF CONNECTION,
THREE:
ONE TO ONE
MANY TO ONE
MANY TO MANY



In [ ]:
from tensorflow.keras.preprocessing.text import Tokenizer
tokenizer=Tokenizer()
tokenizer.fit_on_texts([data])
tokenizer.word_index

{'the': 1,
 'and': 2,
 'of': 3,
 'i': 4,
 'to': 5,
 'a': 6,
 'in': 7,
 'was': 8,
 'it': 9,
 'as': 10,
 'that': 11,
 'my': 12,
 'with': 13,
 'he': 14,
 'on': 15,
 'had': 16,
 'for': 17,
 'which': 18,
 'not': 19,
 'me': 20,
 'but': 21,
 'were': 22,
 'we': 23,
 'his': 24,
 'at': 25,
 'they': 26,
 'you': 27,
 'all': 28,
 'could': 29,
 'there': 30,
 'so': 31,
 'seemed': 32,
 'from': 33,
 'is': 34,
 'by': 35,
 'them': 36,
 'have': 37,
 'this': 38,
 'then': 39,
 'driver': 40,
 'very': 41,
 'be': 42,
 'through': 43,
 'us': 44,
 'some': 45,
 'when': 46,
 'or': 47,
 'time': 48,
 'over': 49,
 'what': 50,
 'horses': 51,
 'an': 52,
 'up': 53,
 'said': 54,
 'into': 55,
 'see': 56,
 'great': 57,
 'out': 58,
 'one': 59,
 'do': 60,
 'here': 61,
 'know': 62,
 'are': 63,
 'must': 64,
 'again': 65,
 'though': 66,
 'been': 67,
 'before': 68,
 'white': 69,
 'down': 70,
 'him': 71,
 'our': 72,
 'count': 73,
 'door': 74,
 'side': 75,
 'now': 76,
 'may': 77,
 'no': 78,
 'long': 79,
 'saw': 80,
 'old': 81,
 'ro

TOKENIZATION-pieces of text
if i want characters no need of []
if i want words use[]-item handling

In [ ]:
from typing import Sequence
input_sequence=[]
for sentence in data.split('\n'):
  tokenized_sentence=tokenizer.texts_to_sequences([sentence])[0]
  for i in range(1,len(tokenized_sentence)):
    input_sequence.append(tokenized_sentence[:i+1])
input_sequence

[[440, 4],
 [301, 441],
 [301, 441, 442],
 [722, 77],
 [722, 77, 199],
 [722, 77, 199, 152],
 [722, 77, 199, 152, 723],
 [722, 77, 199, 152, 723, 25],
 [722, 77, 199, 152, 723, 25, 724],
 [722, 77, 199, 152, 723, 25, 724, 725],
 [722, 77, 199, 152, 723, 25, 724, 725, 726],
 [722, 77, 199, 152, 723, 25, 724, 725, 726, 727],
 [722, 77, 199, 152, 723, 25, 724, 725, 726, 727, 15],
 [722, 77, 199, 152, 723, 25, 724, 725, 726, 727, 15, 728],
 [722, 77, 199, 152, 723, 25, 724, 725, 726, 727, 15, 728, 77],
 [722, 77, 199, 152, 723, 25, 724, 725, 726, 727, 15, 728, 77, 729],
 [722, 77, 199, 152, 723, 25, 724, 725, 726, 727, 15, 728, 77, 729, 25],
 [730, 443],
 [730, 443, 302],
 [730, 443, 302, 200],
 [730, 443, 302, 200, 130],
 [730, 443, 302, 200, 130, 37],
 [730, 443, 302, 200, 130, 37, 444],
 [730, 443, 302, 200, 130, 37, 444, 25],
 [730, 443, 302, 200, 130, 37, 444, 25, 731],
 [730, 443, 302, 200, 130, 37, 444, 25, 731, 732],
 [730, 443, 302, 200, 130, 37, 444, 25, 731, 732, 21],
 [730, 443

text data->sequence i+1 slicing nxt

In [ ]:
max_len=max([ len (x) for x in input_sequence])
max_len


17

converting to ml
[440, 4],need 1
[301, 441],need 1
[301, 441, 442],
pairing method -matchup
->padding=putting 0 (dummy)
 pre and post pdding
 maximum length of matrix


In [ ]:
from tensorflow.keras.preprocessing.sequence import pad_sequences
padd=pad_sequences(input_sequence,maxlen=max_len,padding="pre")
padd

array([[   0,    0,    0, ...,    0,  440,    4],
       [   0,    0,    0, ...,    0,  301,  441],
       [   0,    0,    0, ...,  301,  441,  442],
       ...,
       [   0,    0,    0, ...,    2,    4, 1782],
       [   0,    0,    0, ...,    4, 1782,   12],
       [   0,    0,    0, ..., 1782,   12,  438]], dtype=int32)

In [ ]:
x=padd[:,:-1]
y=padd[:,-1]
y

array([   4,  441,  442, ..., 1782,   12,  438], dtype=int32)

In [ ]:
print(x.shape)
print(y.shape)
print(len(tokenizer.word_index))

(7047, 16)
(7047,)
1782


as of now we are using  softmax->multiclassification


In [ ]:
from tensorflow.keras.utils import to_categorical
y=to_categorical(y,num_classes=1783)

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding,SimpleRNN,Dense
model=Sequential([
    Embedding(1783,100,input_length=16),
    SimpleRNN(150),
    Dense(1783,activation="softmax")
])
model.compile(loss="categorical_crossentropy",optimizer="adam",metrics=["accuracy"])
model.fit(x,y,epochs=20)


Epoch 1/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 8s 22ms/step - accuracy: 0.0595 - loss: 6.6040
Epoch 2/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 4s 16ms/step - accuracy: 0.0739 - loss: 5.8799
Epoch 3/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 4s 16ms/step - accuracy: 0.0876 - loss: 5.5612
Epoch 4/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 6s 19ms/step - accuracy: 0.1042 - loss: 5.2848
Epoch 5/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 4s 16ms/step - accuracy: 0.1256 - loss: 4.9203
Epoch 6/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 4s 16ms/step - accuracy: 0.1518 - loss: 4.5450
Epoch 7/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 5s 22ms/step - accuracy: 0.1838 - loss: 4.2078
Epoch 8/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 4s 16ms/step - accuracy: 0.2466 - loss: 3.8179
Epoch 9/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 4s 16ms/step - accuracy: 0.2944 - loss: 3.4634
Epoch 10/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 5s 22ms/step - accuracy: 0.3760 - loss: 3.1168
Epoch 11/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 4s 16ms/step - accuracy: 0.4544 - loss: 2.7402
Epoch 12/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 

here first you have to call embedding
output dimensional -1783,input dimensional-moderate -100,input length
COMPILE-READY

In [ ]:
import numpy as np
import time
text="where"
for i in range(10):
  ss=tokenizer.texts_to_sequences([text])[0]
  dd=pad_sequences([ss],maxlen=max_len-1,padding="pre")
  pos=np.argmax(model.predict(dd))
  for word,index in tokenizer.word_index.items():
    if pos==index:
      text=text +" "+word
      print(text)
      time.sleep(1)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 63ms/step
where you
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 46ms/step
where you will
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step
where you will need
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step
where you will need to
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step
where you will need to night
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step
where you will need to night at
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step
where you will need to night at least
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step
where you will need to night at least to
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step
where you will need to night at least to wait
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step
where you will need to night at least to wait the


next word prediction
because we are using new data it should be cleaned



In [ ]:
model.save("RNN.h5")

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding,LSTM,Dense
model=Sequential([
    Embedding(1783,100,input_length=16),
    LSTM(150),
    Dense(1783,activation="softmax")
])
model.compile(loss="categorical_crossentropy",optimizer="adam",metrics=["accuracy"])
model.fit(x,y,epochs=20)

Epoch 1/20


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/embedding.py:97: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


221/221 ━━━━━━━━━━━━━━━━━━━━ 11s 40ms/step - accuracy: 0.0613 - loss: 6.6277
Epoch 2/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 9s 34ms/step - accuracy: 0.0691 - loss: 5.8590
Epoch 3/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 10s 35ms/step - accuracy: 0.0717 - loss: 5.7431
Epoch 4/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 11s 41ms/step - accuracy: 0.0915 - loss: 5.5897
Epoch 5/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 10s 40ms/step - accuracy: 0.0882 - loss: 5.5190
Epoch 6/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 9s 40ms/step - accuracy: 0.0949 - loss: 5.3733
Epoch 7/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 8s 34ms/step - accuracy: 0.1112 - loss: 5.2016
Epoch 8/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 10s 47ms/step - accuracy: 0.1307 - loss: 4.9820
Epoch 9/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 19s 39ms/step - accuracy: 0.1531 - loss: 4.7466
Epoch 10/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 8s 36ms/step - accuracy: 0.1502 - loss: 4.6554
Epoch 11/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 9s 41ms/step - accuracy: 0.1776 - loss: 4.3735
Epoch 12/20
221/221 ━━━━━━━━━━━━━━━━━━━━ 10s 4